File: pdf_reader.php
<?php
// pdf_reader.php
// Версия на PHP с использованием smalot/pdfparser

require_once 'vendor/autoload.php';

use Smalot\PdfParser\Parser;

// ANSI-цвета
define('RESET', "\033[0m");
define('CYAN', "\033[96m");
define('GREEN', "\033[92m");
define('YELLOW', "\033[93m");
define('RED', "\033[91m");
define('BLUE', "\033[94m");
define('BOLD', "\033[1m");

function colorize($text, $color) {
    return $color . $text . RESET;
}

class PDFReader {
    private $filename;
    private $password;
    private $parser;
    private $pdf;

    public function __construct($filename, $password = null) {
        $this->filename = $filename;
        $this->password = $password;
        $this->parser = new Parser();
        $this->pdf = $this->parser->parseFile($filename);
        // Пароль не поддерживается напрямую, но оставим для совместимости
    }

    public function getMetadata() {
        $details = $this->pdf->getDetails();
        $meta = [
            'Title' => $details['Title'] ?? '',
            'Author' => $details['Author'] ?? '',
            'Creator' => $details['Creator'] ?? '',
            'Producer' => $details['Producer'] ?? '',
            'CreationDate' => $details['CreationDate'] ?? '',
            'ModDate' => $details['ModDate'] ?? '',
        ];
        return $meta;
    }

    public function getPageCount() {
        return count($this->pdf->getPages());
    }

    public function extractText($pageNum = null) {
        $text = '';
        if ($pageNum !== null) {
            $pages = $this->pdf->getPages();
            if ($pageNum < 1 || $pageNum > count($pages)) {
                throw new Exception("Номер страницы вне диапазона");
            }
            $text = $pages[$pageNum-1]->getText();
        } else {
            $text = $this->pdf->getText();
        }
        return $text;
    }

    public function search($query, $pageNum = null) {
        $text = $this->extractText($pageNum);
        $lines = explode("\n", $text);
        $results = [];
        foreach ($lines as $i => $line) {
            if (stripos($line, $query) !== false) {
                $results[$i+1] = trim($line);
            }
        }
        return $results;
    }
}

// Парсинг аргументов
$shortOpts = "tmps:o:";
$longOpts = ['password:', 'page:'];
$options = getopt($shortOpts, $longOpts);
$args = array_values(array_filter($argv, function($arg) { return !str_starts_with($arg, '-'); }));
array_shift($args);

$filename = $args[0] ?? null;
$password = $options['password'] ?? null;
$pageNum = isset($options['page']) ? (int)$options['page'] : 0;
$textFlag = isset($options['t']);
$metadataFlag = isset($options['m']);
$pagesFlag = isset($options['p']);
$searchQuery = $options['s'] ?? null;
$outputFile = $options['o'] ?? null;

if (!$filename) {
    echo "Укажите PDF файл.\n";
    exit(1);
}

try {
    $reader = new PDFReader($filename, $password);

    if ($metadataFlag) {
        $meta = $reader->getMetadata();
        echo colorize("Метаданные:", BOLD) . "\n";
        foreach ($meta as $k => $v) {
            echo "  " . colorize($k, CYAN) . ": $v\n";
        }
        echo "\n";
    }

    if ($pagesFlag) {
        $count = $reader->getPageCount();
        echo "Количество страниц: " . colorize($count, GREEN) . "\n\n";
    }

    if ($searchQuery) {
        $results = $reader->search($searchQuery, $pageNum);
        if (count($results) > 0) {
            echo colorize("Найдено совпадений: " . count($results), GREEN) . "\n";
            foreach ($results as $lineNum => $line) {
                $highlighted = preg_replace('/' . preg_quote($searchQuery, '/') . '/i', colorize('$0', YELLOW), $line);
                echo "  Строка $lineNum: $highlighted\n";
            }
        } else {
            echo colorize("Совпадений не найдено.", RED) . "\n";
        }
        echo "\n";
    }

    if ($textFlag || (!$metadataFlag && !$pagesFlag && !$searchQuery)) {
        $text = $reader->extractText($pageNum);
        if ($outputFile) {
            file_put_contents($outputFile, $text);
            echo colorize("Текст сохранён в $outputFile", GREEN) . "\n";
        } else {
            echo colorize("Извлечённый текст:", BOLD) . "\n";
            echo $text;
        }
    }
} catch (Exception $e) {
    echo colorize("Ошибка: " . $e->getMessage(), RED) . "\n";
    exit(1);
}

File: pdf_reader.go
// pdf_reader.go
// Версия на Go с использованием github.com/unidoc/unidoc (или ledongthuc/pdf)
// Для упрощения используем ledongthuc/pdf (легче в установке)

package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/ledongthuc/pdf"
)

// ANSI-цвета
const (
	reset  = "\033[0m"
	cyan   = "\033[96m"
	green  = "\033[92m"
	yellow = "\033[93m"
	red    = "\033[91m"
	blue   = "\033[94m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type PDFReader struct {
	filename string
	password string
}

func NewPDFReader(filename, password string) *PDFReader {
	return &PDFReader{filename: filename, password: password}
}

func (r *PDFReader) Open() (*pdf.Reader, error) {
	f, err := os.Open(r.filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	reader, err := pdf.NewReader(f, r.password)
	if err != nil {
		return nil, err
	}
	return reader, nil
}

func (r *PDFReader) GetMetadata() (map[string]string, error) {
	reader, err := r.Open()
	if err != nil {
		return nil, err
	}
	meta := reader.Trailer()
	result := make(map[string]string)
	if meta != nil {
		for _, key := range []string{"Title", "Author", "Creator", "Producer", "CreationDate", "ModDate"} {
			if val := meta.Key(key).String(); val != "" {
				result[key] = val
			}
		}
	}
	return result, nil
}

func (r *PDFReader) GetPageCount() (int, error) {
	reader, err := r.Open()
	if err != nil {
		return 0, err
	}
	return reader.NumPage(), nil
}

func (r *PDFReader) ExtractText(pageNum int) (string, error) {
	reader, err := r.Open()
	if err != nil {
		return "", err
	}
	var builder strings.Builder
	if pageNum > 0 {
		if pageNum < 1 || pageNum > reader.NumPage() {
			return "", fmt.Errorf("номер страницы вне диапазона")
		}
		page := reader.Page(pageNum)
		text, err := page.GetPlainText(nil)
		if err != nil {
			return "", err
		}
		builder.WriteString(text)
	} else {
		for i := 1; i <= reader.NumPage(); i++ {
			page := reader.Page(i)
			text, err := page.GetPlainText(nil)
			if err != nil {
				return "", err
			}
			builder.WriteString(text)
		}
	}
	return builder.String(), nil
}

func (r *PDFReader) Search(query string, pageNum int) (map[int][]string, error) {
	text, err := r.ExtractText(pageNum)
	if err != nil {
		return nil, err
	}
	lines := strings.Split(text, "\n")
	results := make(map[int][]string)
	for i, line := range lines {
		if strings.Contains(strings.ToLower(line), strings.ToLower(query)) {
			results[i+1] = append(results[i+1], strings.TrimSpace(line))
		}
	}
	return results, nil
}

func main() {
	var (
		textFlag      bool
		metadataFlag  bool
		pagesFlag     bool
		searchQuery   string
		outputFile    string
		password      string
		pageNum       int
	)
	flag.BoolVar(&textFlag, "t", false, "Извлечь текст")
	flag.BoolVar(&metadataFlag, "m", false, "Показать метаданные")
	flag.BoolVar(&pagesFlag, "p", false, "Показать количество страниц")
	flag.StringVar(&searchQuery, "s", "", "Искать текст")
	flag.StringVar(&outputFile, "o", "", "Сохранить текст в файл")
	flag.StringVar(&password, "password", "", "Пароль для PDF")
	flag.IntVar(&pageNum, "page", 0, "Номер страницы (1-based)")
	flag.Usage = func() {
		fmt.Println("Использование: go run pdf_reader.go [опции] <pdf_file>")
		fmt.Println("  -t            Извлечь текст")
		fmt.Println("  -m            Показать метаданные")
		fmt.Println("  -p            Показать количество страниц")
		fmt.Println("  -s <query>    Искать текст")
		fmt.Println("  -o <file>     Сохранить текст в файл")
		fmt.Println("  --password    Пароль для PDF")
		fmt.Println("  --page <num>  Номер страницы")
	}
	flag.Parse()

	if flag.NArg() < 1 {
		flag.Usage()
		os.Exit(1)
	}
	filename := flag.Arg(0)

	reader := NewPDFReader(filename, password)

	if metadataFlag {
		meta, err := reader.GetMetadata()
		if err != nil {
			fmt.Println(colorize("Ошибка: "+err.Error(), red))
			os.Exit(1)
		}
		fmt.Println(colorize("Метаданные:", bold))
		for k, v := range meta {
			fmt.Printf("  %s: %s\n", colorize(k, cyan), v)
		}
		fmt.Println()
	}

	if pagesFlag {
		count, err := reader.GetPageCount()
		if err != nil {
			fmt.Println(colorize("Ошибка: "+err.Error(), red))
			os.Exit(1)
		}
		fmt.Printf("Количество страниц: %s\n", colorize(fmt.Sprintf("%d", count), green))
		fmt.Println()
	}

	if searchQuery != "" {
		results, err := reader.Search(searchQuery, pageNum)
		if err != nil {
			fmt.Println(colorize("Ошибка: "+err.Error(), red))
			os.Exit(1)
		}
		if len(results) > 0 {
			fmt.Printf(colorize("Найдено совпадений: %d\n", green), len(results))
			for lineNum, lines := range results {
				for _, line := range lines {
					highlighted := strings.ReplaceAll(line, searchQuery, colorize(searchQuery, yellow))
					fmt.Printf("  Строка %d: %s\n", lineNum, highlighted)
				}
			}
		} else {
			fmt.Println(colorize("Совпадений не найдено.", red))
		}
		fmt.Println()
	}

	if textFlag || (!metadataFlag && !pagesFlag && searchQuery == "") {
		text, err := reader.ExtractText(pageNum)
		if err != nil {
			fmt.Println(colorize("Ошибка: "+err.Error(), red))
			os.Exit(1)
		}
		if outputFile != "" {
			err = os.WriteFile(outputFile, []byte(text), 0644)
			if err != nil {
				fmt.Println(colorize("Ошибка записи: "+err.Error(), red))
				os.Exit(1)
			}
			fmt.Println(colorize("Текст сохранён в "+outputFile, green))
		} else {
			fmt.Println(colorize("Извлечённый текст:", bold))
			fmt.Println(text)
		}
	}
}
